staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go

```diff
@@ -34,13 +35,102 @@ import (
 	apiservervalidation "k8s.io/apiextensions-apiserver/pkg/apiserver/validation"
 )
 
+type CustomResourceValidator interface {
+	Validate(ctx context.Context, obj runtime.Object, scale *apiextensions.CustomResourceSubresourceScale) field.ErrorList
+	ValidateUpdate(ctx context.Context, obj, old runtime.Object, scale *apiextensions.CustomResourceSubresourceScale) field.ErrorList
+	ValidateStatusUpdate(ctx context.Context, obj, old runtime.Object, scale *apiextensions.CustomResourceSubresourceScale) field.ErrorList
+}
+
+type lazyValidator struct {
+	newSchemaValidator, newStatusSchemaValidator func() (*validate.SchemaValidator, error)
+
+	lock      sync.RWMutex
+	validator *customResourceValidator
+}
+
+func NewLazyValidator(namespaceScoped bool, kind schema.GroupVersionKind, newSchemaValidator, newStatusSchemaValidator func() (*validate.SchemaValidator, error)) CustomResourceValidator {
+	return &lazyValidator{
+		newSchemaValidator:       newSchemaValidator,
+		newStatusSchemaValidator: newStatusSchemaValidator,
+		validator: &customResourceValidator{
+			namespaceScoped: namespaceScoped,
+			kind:            kind,
+		},
+	}
+}
+
+func (v *lazyValidator) Validate(ctx context.Context, obj runtime.Object, scale *apiextensions.CustomResourceSubresourceScale) field.ErrorList {
+	v.lock.RLock()
+	if v.validator.schemaValidator == nil {
+		v.lock.RUnlock()
+		v.lock.Lock()
```

Comment:
the lock logic is really complex, isn't it?

Reply:
Is it? It's the standard way to first read lock, then write lock on fail, check again and do the slow path. Am open for a better way to write that.

Reply:
I just was wondering if there is really improvement between locking the whole operation vs doing this intermediate locks switching from RLock to Lock

Reply:
Both operations are expensive, and on a critical path of requests. Hence, I would rather keep this additional complexity.

Reply:
This is another example of `sync.Once` if I'm reading correctly.

EDIT: Actually no because you might want to regenerate if it fails. I don't see the actual benefits of using a RW lock here to be honest since you're only protecting the pointer (and checking if the pointer is nil). the contention is minimal unless proven otherwise. The only thing that will hold the lock longer is holding a write lock which is going to block everything anyway.

Reply:
`newSchemaValidator` is heavy, `v.validator.Validate` is heavy. Typical case for a RW lock.

It might be a discussion about which risk of contention is enough to give reason to use a RWLock. But that's philosophical.

Reply:
I think a static source would do here.